### DISCUSSION TITLE
Pity state of packaging in distro(s)

### DISCUSSION THREAD

**Discussion by crosser** [svinota/pyroute2#884] [Q&A]
We use debian as the base for our production systems, and I used to make `.deb` packages from previous versions of pyroute2 for internal use, when we could not afford to wait for the new version to arrive in debian. It used to be pretty straightforward.

Now it's the first time that I need to do it again since the library was split in several parts. My old debianisation would not work, and I went upstream to borrow it from there. I found that they butchered ~pr2modules/config/version.py~ `util/update_version.py`, introduced build dependency on openstack(!) and use some openstack env variable to establish version. Basically, build-dependency on openstack prevents me from using this packaging configuration for our system.

Still, there is a real pain point here: Makefile tries to make VERSION, and it requires one to build from live git repo. An "accepted" way to execute a (at least somewhat reproducible) package build is to have a clean source tree at the start (called `something_orig.tar.gz` in debian) with `VERSION` and `pyroute2.core/pr2modules/config/version.py` already created, that `make` would not suddenly try to re-create.

So, is there a way to produce a .git-free source tree (tar), that

1. Includes `VERSION` and `pyroute2.core/pr2modules/config/version.py`
2. Has the logic of re-creating these files _disabled_

If there is no procedure to do that, should it be introduced? Or am I missing the point altogether, and there is a different way to cleanly build packages?

Thanks!

**Comment by svinota** [comment]
I could add version files to the tree and update them with every new tag, if it can help you with packaging.

**Reply by crosser** [reply]
Another option would be to have a target a la `make source-tar` that would produce pristine "orig" tar file with version files included in it.

But! in either case, there is a danger that the user (or git) will mess with timestamps, and `make` will decide to re-create them. And fail because it won't be able to run `git describe`.

I don't really know a "right" solution. Maybe "de-fang" `util/update_version.py` script, so it does not fail if (1) there are version files, and (2) it cannot run `git describe`?

Something along these lines:
```
diff --git a/util/update_version.py b/util/update_version.py
index a71be3f1..f240e384 100644
--- a/util/update_version.py
+++ b/util/update_version.py
@@ -1,4 +1,5 @@
 #!/usr/bin/env python
+from os.path import exists
 import subprocess
 
 version_module = 'pyroute2.core/pr2modules/config/version.py'
@@ -27,7 +28,9 @@ def get_project_version():
 
 if __name__ == '__main__':
     version = get_project_version()
-    with open(version_module, 'w') as f:
-        f.write('__version__ = "%s"\n' % version)
-    with open(version_file, 'w') as f:
-        f.write('%s\n' % version)
+    if not (version == 'unknown' and exists(version_module)
+            and exists(version_file)):
+        with open(version_module, 'w') as f:
+            f.write('__version__ = "%s"\n' % version)
+        with open(version_file, 'w') as f:
+            f.write('%s\n' % version)
```

**Reply by crosser** [reply]
It's actually quite a mess, because `make clean` removes these files, etc. If they are not removed, then the target `VERSION` is bypassed and additional copying in that target is not done. Etc.

Looks like I made things work somehow for me; fixing them properly would be too tedious :frowning_face:

**Reply by svinota** [reply]
I planned tonight to provide static version files, tell me if it will break your scheme again.

So far the plan is:

1. To have static version files
2. Update the version files with every new tag
3. Generate versions like `x.x.x-x` only if requested; don't commit these changes, don't accept them in PR:s
4. (Completely deprecate nose tests) ⇐ side task
5. Test the files with every CI run

**Reply by crosser** [reply]
I hope that I will be able to test your changes.

A sidenote: `pyroute2.minimal/setup.py` is in the repo, but it is covered by a wildcard in `.gitignore`, this is probably wrong...

**Reply by svinota** [reply]
… but there is one thing that makes me hesitating

if we leave static version files, then it will be inevitable that two different deployments from the master branch on different commits will end up having same version files and that makes me uncomfortable

**Reply by svinota** [reply]
https://github.com/svinota/pyroute2/commit/951b1d0e478419ca7cf57768a06e4896fa0946eb

Committed a solution that may fit. Pls review

**Reply by crosser** [reply]
> Update the version files with every new tag

In the times of CVS and SVN there was a very useful feature to update a string inside source files on every commit. I was actually baffled when I learned that git has no plans to implement anything similar. Update on every commit would be _the_ right way to handle it. With git, while this _can_ be done locally, it is not part of the repo, so when other people clone the repo and start to make their own changes, information in the files will stop being updated, resulting in stale version info, as you mentioned.

It would seem that the only _consistent_ way is to make builds two-stage - (1) to make a snapshot of the source from git, and (2) use this snapshot to build pip / distutils / deb / rpm package. The second stage should not rely on the presence of git.

As you mentioned, in the current approach, same version number will be assigned to different versions of the source...

Also, how will it work from technical perspective? Are you going to first modify the file, commit the change, and then tag? Otherwise tag will be pointing to the state of the tree where `version` still has previous contents...

**Reply by svinota** [reply]
So far the plan is to:

1. update the changelog
2. update the version
3. commit
4. tag
5. ⇒ check it on every CI run

Updating the `version` file on every commit with a pre-commit hook is a little bit overkill, I believe.

**Comment by crosser** [comment]
The current state (fc4990e) makes build nice and clean _for me_, i.e. I don't need any modifications on top the the snapshot of the source code from git, only the added `debian` directory.

But I envision that the taken approach may break for people who build the package from a git repo that is _not_ a clone of the upstream, but repopulated/updated from a snapshot of the upstream code. They will get irrelevant version data. (Maybe there are no such people, and this is not a problem.)

**Reply by svinota** [reply]
I hope that they will let me know if they get any issues.

Regarding your build — good to hear :) Pls don't hesitate to ping me if I can help more.
